packages/data/src/components/use-select/index.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * WordPress dependencies
+ */
+import { createQueue } from '@wordpress/priority-queue';
+import {
+	useLayoutEffect,
+	useEffect,
+	useState,
+	useRef,
+	useMemo,
+	useCallback,
+} from '@wordpress/element';
+import { isShallowEqualObjects } from '@wordpress/is-shallow-equal';
+
+/**
+ * Internal dependencies
+ */
+import useRegistry from '../registry-provider/use-registry';
+import useAsyncMode from '../async-mode-provider/use-async-mode';
+
+/**
+ * Favor useLayoutEffect to ensure the store subscription callback always has
+ * the selector from the latest render. If a store update happens between render
+ * and the effect, this could cause missed/stale updates or inconsistent state.
+ *
+ * Fallback to useEffect for server rendered components because currently React
+ * throws a warning when using useLayoutEffect in that environment.
+ */
+// const useIsomorphicLayoutEffect = typeof window !== 'undefined' ?
+// 	useLayoutEffect : useEffect;
+const useIsomorphicLayoutEffect = typeof window !== 'undefined' ?
+	useLayoutEffect : useEffect;
+
+const renderQueue = createQueue();
+
+export default function useSelect( mapSelect, nextProps ) {
+	const registry = useRegistry();
+	const isAsync = useAsyncMode();
+	const [ mapOutput, setMapOutput ] = useState( mapSelect( registry.select ) );
+	const queueContext = useMemo( () => ( { queue: true } ), [ registry ] );
+
+	const latestMapSelect = useRef( mapSelect );
+	const latestIsAsync = useRef( isAsync );
+	const latestMapOutput = useRef( mapOutput );
```

Comment:
It's not clear for me (yet) why we need all these refs.

Reply:
Ya I'm still on the fence with these as well.  

- for `latestMapSelect` its to avoid rerenders on any mapSelect change but still ensure we're using the latest.  This possibly reduces the need for implementors to provide a memoized (via `useCallback` mapSelect.
- `latestAsync` ya this one is unnecessary because `isAsync` is a dependency on the effects.
- `latestMapOutput` is needed to compare the previous mapOutput with the latest generated to only return the latest if it's changed.  Alternatively we might be able to use `useMemo` for this instead.

Reply:
> latestAsync ya this one is unnecessary because isAsync is a dependency on the effects.

Why not just use `isAsync` directly?

Reply:
> Why not just use isAsync directly?

That's essentially what I'm meaning, since `isAsync` is used as a dependency I can just use it directly.  So there is no need for `useRef`.

Reply:
sorry I misread :)

Reply:
Hmm I see why I am using `useRef` now.  The effect that is setting the subscription only sets/unsets the subscription if the registry changes, but it the callback will conditionally use the render queue if `isAsync` is true.  To avoid making the subscription callback dependent on `isAsync` changes (i.e. the effect getting invoked if `isAsync` changes) and keep its only dependency as the registry, the `useRef` is used to make sure that `isAsync` is always the latest value.

Reply:
you don't have to add `isAsync` to the dependencies of that effect right?

Reply:
We need all of these to be refs to avoid resubscribing on every render.

Reply:
we only resubscribe on every render if we add explicitly the dependency to the effect. So instead of using "refs" and adding them as dependencies, why not just use the props directly and omitting them from the dependencies.

Reply:
> instead of using "refs" and adding them as dependencies,

I'm not sure what you mean here.  The refs are not added as dependencies... that's the point?

Reply:
So, I think I understand better now :) 

I think there's a bug somewhere. When `isAsync` changes, in addition to the "flusing", we need to change the subscription model (retrigger what you do right now when the registry changes) and at that point, there's no reason to keep `isAsync` as ref.

Reply:
Yeah, because they are closures, you either need to resubscribe every time they change or use refs.

>retrigger what you do right now when the registry changes

That already happens.

Reply:
yes but not when isAsync changes and it should at that point.

Reply:
It doesn't need to, because `isAsync` is referenced through a ref.

Reply:
> yes but not when isAsync changes and it should at that point.

If you're talking about the subscribe changing when `isAsync` changes... I'm not sure I see that in the old `withSelect` as a reference point.  The only time I see the subscription changing is when the registry changes:

https://github.com/WordPress/gutenberg/blob/ba983324239faa5b6628034be9aab2c901aecb48/packages/data/src/components/with-select/with-select-old.js#L111-L114

Reply:
Mmm you're absolutely right @nerrad I forgot that it was handled in the callback and not at the subscription level.